src/security/ServerOptions.cc

```diff
@@ -372,9 +376,36 @@ Security::ServerOptions::loadDhParams()
             dhp = nullptr;
         }
     }
-
     parsedDhParams.resetWithoutLocking(dhp);
+
+#else // OpenSSL 3.0+
+    EVP_PKEY *pkey = nullptr;
+    const char *type = "DH";
+    if (!eecdhCurve.isEmpty())
+        type = "EC";
+
+    if (auto *dctx = OSSL_DECODER_CTX_new_for_pkey(&pkey, "PEM", nullptr, type, 0, nullptr, nullptr)) {
+        if (auto *in = fopen(dhParamsFile.c_str(), "r")) {
+            if (OSSL_DECODER_from_fp(dctx, in) == 1) {
+
+                /* pkey is created with the decoded data from the bio */
+                Must(pkey);
+                parsedDhParams.resetWithoutLocking(pkey);
+                // TODO: verify that the loaded parameters match the value in eecdhCurve
```

Comment:
The official code uses DH_check() for that. Can we just do the same (or something similar) here? Sorry if you have already explained why this TODO is not implemented in this PR, breaking OpenSSL v1 feature parity -- please feel free to refer me to that explanation.

Reply:
Please do.

What needs ding is verifying that parsedDhParams object contains EC parameters that map to the Curve named in eecdhCurve string.

Last I looked there was a complications:
 * some parameters have multiple Curve names. So a reverse mapping check may not result in the original name.
 *  it is valid to load a file which has custom parameters that do not map to any of the well-known Curve names. So we may need to define a custom eecdhCurve string to skip the check for these edge cases.

Reply:
OK, this change request is on my side.

Reply:
Status update: I have the validation code for DH parameters, but there is a problem with the surrounding code that should be fixed before the validation code can run. I am polishing that fix as well and hope to commit all the changes in 24-48 hours.

Reply:
> Alex: we should at least try to preserve the feature parity with OpenSSL v1 in this context.

Done in e794b46. I am resolving this change request. While working on this changes, I discovered and fixed a few related problems. Please see commit messages for details, especially 75bd3aa and f2830c5.

> Amos: What needs doing is verifying that parsedDhParams object contains EC parameters that map to the Curve named in eecdhCurve string

AFAICT, all of that (and associated complications) are outside this PR scope because OpenSSL v1 code was not verifying any of that either.